File: app/src/main/assets/modules/__base64__.js
/**
 * @param {ScriptRuntime} scriptRuntime
 * @param {org.mozilla.javascript.Scriptable | global} scope
 * @return {Internal.Base64}
 */
module.exports = function (scriptRuntime, scope) {
    const Base64 = android.util.Base64;
    let _ = {
        Base64Ctor: (/* @IIFE */ () => {
            /**
             * @implements Internal.Base64
             */
            const Base64Ctor = function () {
                /* Empty body. */
            };

            Base64Ctor.prototype = {
                constructor: Base64Ctor,
                encode(o, encoding) {
                    let niceEncoding = _.parseEncoding(encoding);
                    return Base64.encodeToString(_.toBytes(o, niceEncoding), Base64.NO_WRAP);
                },
                decode(o, encoding) {
                    let niceEncoding = _.parseEncoding(encoding);
                    let decoded = Base64.decode(_.toBytes(o, niceEncoding), Base64.NO_WRAP);
                    // noinspection JSValidateTypes
                    return typeof niceEncoding !== 'undefined'
                        ? String(new java.lang.String(decoded, niceEncoding))
                        : String(new java.lang.String(decoded));
                },
            };

            return Base64Ctor;
        })(),
        /**
         * Ignored (but not fuzzy) regex matching for non-word characters.
         */
        parseEncoding: (encoding) => {
            return [
                StandardCharsets.ISO_8859_1,
                StandardCharsets.US_ASCII,
                StandardCharsets.UTF_8,
                StandardCharsets.UTF_16,
                StandardCharsets.UTF_16BE,
                StandardCharsets.UTF_16LE,
            ].find((cs) => cs.name().toLowerCase().replace(/\W+/g, '')
                === String(encoding).trim().toLowerCase().replace(/\W+/g, ''));
        },
        toBytes(o, encoding) {
            if (typeof o === 'string') {
                // noinspection JSValidateTypes
                /** @type { java.lang.String } */
                let string = new java.lang.String(o);
                return typeof encoding !== 'undefined'
                    ? string.getBytes(encoding)
                    : string.getBytes();
            }
            if (util.getClassName(o) === '[B') {
                return o;
            }
            if (Array.isArray(o)) {
                return ArrayUtils.jsBytesToByteArray(o);
            }
            if (!isNullish(o)) {
                if (typeof o.toString === 'function') {
                    return this.toBytes(o.toString());
                }
            }
            throw Error(`Can't convert o (${o}) to bytes`);
        },
    };

    // noinspection UnnecessaryLocalVariableJS
    /**
     * @type {Internal.Base64}
     */
    const base64 = new _.Base64Ctor();

    return base64;
};

File: app/src/main/java/org/autojs/autojs/runtime/api/augment/base64/Base64.kt
package org.autojs.autojs.runtime.api.augment.base64

import org.autojs.autojs.annotation.RhinoFunctionBody
import org.autojs.autojs.annotation.RhinoSingletonFunctionInterface
import org.autojs.autojs.extension.AnyExtensions.isJsNullish
import org.autojs.autojs.runtime.api.augment.Augmentable
import org.autojs.autojs.runtime.exception.WrappedIllegalArgumentException
import org.autojs.autojs.util.ArrayUtils
import org.mozilla.javascript.Context
import org.mozilla.javascript.NativeArray
import java.nio.charset.Charset
import android.util.Base64 as AndroidBase64

object Base64 : Augmentable() {

    override val selfAssignmentFunctions = listOf(
        ::encode.name,
        ::decode.name,
    )

    @JvmStatic
    @RhinoSingletonFunctionInterface
    fun encode(args: Array<out Any?>): String = ensureArgumentsLengthInRange(args, 1..2) {
        val (o, encoding) = it
        encodeRhino(o, encoding)
    }

    @JvmStatic
    @RhinoFunctionBody
    fun encodeRhino(o: Any?, encoding: Any?): String {
        return AndroidBase64.encodeToString(toBytes(o, parseEncoding(encoding)), AndroidBase64.NO_WRAP)
    }

    @JvmStatic
    @RhinoSingletonFunctionInterface
    fun decode(args: Array<out Any?>): String = ensureArgumentsLengthInRange(args, 1..2) {
        val (o, encoding) = it
        decodeRhino(o, encoding)
    }

    @JvmStatic
    @RhinoFunctionBody
    fun decodeRhino(o: Any?, encoding: Any?): String {
        val niceEncoding = parseEncoding(encoding)
        val decoded = AndroidBase64.decode(toBytes(o, niceEncoding), AndroidBase64.NO_WRAP)
        return String(decoded, niceEncoding)
    }

    /**
     * Ignored (but not fuzzy) regex matching for non-word characters.
     */
    private fun parseEncoding(encoding: Any?): Charset {
        if (encoding.isJsNullish()) return Charset.defaultCharset()
        val niceEncoding = Context.toString(encoding)
        val charsetEncoding = niceEncoding.trim().lowercase().replace("\\W+".toRegex(), "")
        return Charset.availableCharsets().values.find {
            it.name().lowercase().replace("\\W+".toRegex(), "") == charsetEncoding
        } ?: Charset.defaultCharset()
    }

    private fun toBytes(o: Any?, encoding: Charset = Charset.defaultCharset()): ByteArray = when {
        o.isJsNullish() -> throw WrappedIllegalArgumentException("Cannot convert o ($o) to byte[]")
        o is String -> o.toByteArray(encoding)
        o is ByteArray -> o
        o is NativeArray -> ArrayUtils.jsBytesToByteArray(o)
        else -> toBytes(o.toString())
    }

}